types/priv_validator/socket.go

```diff
@@ -71,148 +90,276 @@ func (pvsc *PrivValidatorSocketClient) OnStop() {
 	if err != nil {
 		return err
 	}
-	*vote = *res.(SignVoteMsg).Vote
+
+	*vote = *res.(*SignVoteMsg).Vote
+
 	return nil
 }
 
+// SignProposal implements PrivValidator.
 func (pvsc *PrivValidatorSocketClient) SignProposal(chainID string, proposal *types.Proposal) error {
-	res, err := readWrite(pvsc.conn, SignProposalMsg{Proposal: proposal})
+	err := writeMsg(pvsc.conn, &SignProposalMsg{Proposal: proposal})
+	if err != nil {
+		return err
+	}
+
+	res, err := readMsg(pvsc.conn)
 	if err != nil {
 		return err
 	}
-	*proposal = *res.(SignProposalMsg).Proposal
+
+	*proposal = *res.(*SignProposalMsg).Proposal
+
 	return nil
 }
 
+// SignHeartbeat implements PrivValidator.
 func (pvsc *PrivValidatorSocketClient) SignHeartbeat(chainID string, heartbeat *types.Heartbeat) error {
-	res, err := readWrite(pvsc.conn, SignHeartbeatMsg{Heartbeat: heartbeat})
+	err := writeMsg(pvsc.conn, &SignHeartbeatMsg{Heartbeat: heartbeat})
+	if err != nil {
+		return err
+	}
+
+	res, err := readMsg(pvsc.conn)
 	if err != nil {
 		return err
 	}
-	*heartbeat = *res.(SignHeartbeatMsg).Heartbeat
+
+	*heartbeat = *res.(*SignHeartbeatMsg).Heartbeat
+
 	return nil
 }
 
+func (pvsc *PrivValidatorSocketClient) connect() (net.Conn, error) {
+	retries := dialRetryMax
+
+RETRY_LOOP:
+	for retries > 0 {
+		if retries != dialRetryMax {
+			time.Sleep(time.Second * dialRetryIntervalSeconds)
+		}
+
+		retries--
+
+		conn, err := cmn.Connect(pvsc.SocketAddress)
+		if err != nil {
+			pvsc.Logger.Error(
+				"pvsc connect",
+				"addr", pvsc.SocketAddress,
+				"err", errors.Wrap(err, "connection failed"),
+			)
+
+			continue RETRY_LOOP
+		}
+
+		if err := conn.SetDeadline(time.Now().Add(connDeadline)); err != nil {
+			pvsc.Logger.Error(
+				"pvsc connect",
+				"err", errors.Wrap(err, "setting connection timeout failed"),
+			)
+			continue
+		}
+
+		if pvsc.privKey != nil {
+			conn, err = p2pconn.MakeSecretConnection(conn, pvsc.privKey.Wrap())
+			if err != nil {
+				pvsc.Logger.Error(
+					"pvsc connect",
+					"err", errors.Wrap(err, "encrypting connection failed"),
+				)
+
+				continue RETRY_LOOP
+			}
+		}
+
+		return conn, nil
+	}
+
+	return nil, ErrDialRetryMax
+}
+
 //---------------------------------------------------------
 
 // PrivValidatorSocketServer implements PrivValidator.
 // It responds to requests over a socket
 type PrivValidatorSocketServer struct {
 	cmn.BaseService
 
-	conn        net.Conn
-	proto, addr string
-	listener    net.Listener
+	proto, addr    string
+	listener       net.Listener
+	maxConnections int
+	privKey        *crypto.PrivKeyEd25519
 
 	privVal PrivValidator
 	chainID string
 }
 
-func NewPrivValidatorSocketServer(logger log.Logger, socketAddr, chainID string, privVal PrivValidator) *PrivValidatorSocketServer {
+// NewPrivValidatorSocketServer returns an instance of
+// PrivValidatorSocketServer.
+func NewPrivValidatorSocketServer(
+	logger log.Logger,
+	chainID, socketAddr string,
+	maxConnections int,
+	privVal PrivValidator,
```

Comment:
if SocketServer is what runs in the Tendermint process (ie. it's listening on the socket for a connection from a priv_val_client, but it's also the one that initiates all requests), then it won't have access to a PrivValidator because it doesn't have a key for it - it only gets access to the key by sending messages over the socket